docs/DeveloperGuide.md

```diff
@@ -914,3 +915,24 @@ testers are expected to do more *exploratory* testing.
 
     1. CS2100 TA will be able to view the number of students, number of classes, number of labs.
     2. CS2100 TA will also be able to see the number of labs left to mark.
+  
+  
+## <a name="Effort"></a> **18. Appendix: Effort**
+  
+### Challenges Faced and Achievements Made
+The following is a non-exhaustive list of challenges we encountered during the development of ProgrammerError:
+  1. **Rewriting the Command Syntax** 
+  - This was challenging as it required a nuanced understanding of how CLI syntax is constructed and following standard conventions. 
+  - Not all of us were familiar with Unix syntax, which made it more difficult to design a solution that would be suitable for users who prefer typing.
+  2. **Designing New UI components**
+  - AB3 uses a simple components with an (arguably) poor user interface (UI) and user experience (UX). On the other hand, our team revamped the look and feel of the application, customizing the dark theme specially (using GitHub's color scheme) to suit TAs. This involved much experimentation with color schemes and building new components, such as popup windows to enhance the user experience. 
+  - Other examples include: color-coding the labs which are marked or unmarked (displayed as green or red respectively), as well as customizing the Help window and Dashboard Window to incorporate our new Dark Theme.
+  3. **Working with Third-Party Libraries**
+  - Compared to AB3 which does not use minimal third-party libraries for its features, our team integrated two additional external libraries to develop our `upload` and `download` features. 
+  - This was more challenging than it initally seemed as it required trying out multiple potential candidate libraries to find those which are most suitable for our use case and would be compatible across multiple operating systems. 
+  - Through this process, we also encountered dependency issues with the new libraries and spent significant time and effort debugging these issues.
+  4. **Implementing Dynamic Features**
+  - AB3's features typically work by having the user type in text commands to be executed. We felt this was core, but we wanted to go a step further and implement our features to minimise the TA having to type unecessary commands.
+  - For instance, the list of students always remains sorted whenever updates are made to the student data. This required many checks and careful organization of the student data so that the sorting can be done efficiently at minimal cost to the user TA.
+  - Additionally, the dashboard feature was also designed to update dynamically whenever changes are made to the student data. This saves the TA much time in keeping track of the labs they have yet to mark, thereby enhancing their user experience and satisfaction.
+
```

Comment:
Maybe can also add how the show panel automatically refreshes for any changes made to a student or lab.

Reply:
Fix in latest commit